hazelcast/src/main/java/com/hazelcast/map/impl/query/QueryRunner.java

```diff
@@ -68,22 +70,24 @@ public QueryRunner(MapServiceContext mapServiceContext, QueryOptimizer optimizer
         this.resultProcessorRegistry = resultProcessorRegistry;
     }
 
+    // MIGRATION SAFE QUERYING -> MIGRATION STAMPS ARE VALIDATED (does not have to run on a partition thread)
     // full query = index query (if possible), then partition-scan query
     public Result runIndexOrPartitionScanQueryOnOwnedPartitions(Query query)
             throws ExecutionException, InterruptedException {
-
         int migrationStamp = getMigrationStamp();
         Collection<Integer> initialPartitions = mapServiceContext.getOwnedPartitions();
-
         MapContainer mapContainer = mapServiceContext.getMapContainer(query.getMapName());
 
+        // to optimize the query we need to get any index instance
+        Indexes indexes = mapContainer.getIndexes();
```

Comment:
My thinking was that we avoid allocating the iterator for the path, which is most common (and fastest).